### ISSUE TITLE
[BUG] Videos are not generated when reusing a single page between tests

### ISSUE THREAD

**Issue by ryanrosello-og** [microsoft/playwright#14813]
**Context:**
- Playwright Version:  1.22.2
- Operating System: Windows 11
- Node.js version: v 17.7.1
- Browser: chromium
- Extra: [any specific details about your environment]

## System:
 - OS: Windows 10 10.0.22000
 - Memory: 1.24 GB / 15.79 GB
## Binaries:
 - Node: 17.7.1 - C:\Program Files\nodejs\node.EXE
 - Yarn: 1.22.17 - C:\Program Files\nodejs\yarn.CMD
 - npm: 8.5.2 - C:\Program Files\nodejs\npm.CMD
## Languages:
 - Bash: 5.0.17 - C:\Windows\system32\bash.EXE
 

**Code Snippet**
**playwright.config.ts**

```javascript
const config: PlaywrightTestConfig = {
  reporter: [['dot'],['html']],
  use: {
    trace: 'retain-on-failure',
    screenshot: 'only-on-failure',
    video:'retain-on-failure',
  },
};

export default config;
```

**broke.spec.ts**
(sourced from https://playwright.dev/docs/test-retries#reuse-single-page-between-tests)

```javascript
import { test, Page } from '@playwright/test';

test.describe.configure({ mode: 'serial' });

let page: Page;

test.beforeAll(async ({ browser }) => {
  page = await browser.newPage();
});

test.afterAll(async () => {
  await page.close();
});

test('a failing test', async () => {
  await page.click('text=Get Started fail', {timeout: 1000});  // << deliberate fail
});
```

**Describe the bug**

Videos are not generated when reusing a single page between tests

Set the following configuration:
```javascript
  use: {
    trace: 'retain-on-failure',
    screenshot: 'only-on-failure',
    video:'retain-on-failure',
  },
```
Run a failing test

**Expected Result:**

The spec above should successfully generate a screenshot, test steps, trace file and a video within the HTML result.

**Actual Result:**

The video is not being generated:

![2022-06-13_09-09-29](https://user-images.githubusercontent.com/50574915/173257532-da639abd-ec14-4901-996d-ed8e5af25f7e.png)

**Comment by akarsh17** [comment]
Hi @ryanrosello-og 
If you are picking the browser context configuration directly from xyz.config.js file and have not defined the context in the spec itself, then you would want to try the following configuration in order to get the videos generated. 

  use: {
    trace: 'retain-on-failure',
    screenshot: 'only-on-failure',
    video:'retain-on-failure',
    **contextOptions: {recordVideo: { dir: "\<path_to_store_videos\>/videos/'"}}**
},

**Comment by ryanrosello-og** [comment]
thanks for the suggestion @akarsh17 

I tried the following:

Declared the `recordVideo` directly into the config

```javascript
const config: PlaywrightTestConfig = {
  reporter: [['dot'],['html']],
  use: {
    trace:'on',
    contextOptions : {
      recordVideo:{
        dir:'./videos'
      }
    }
  }
};

export default config;
```

And also directly into the `beforeAll` of the spec:


```javascript
import { test, Page, chromium, Browser } from "@playwright/test";

let page: Page;
let browser: Browser;
test.beforeAll(async () => {
  browser = await chromium.launch();
  const context = await browser.newContext({
    recordVideo: {
      dir: `./videos/`,
    },
  });
  page = await context.newPage();
});

test.afterAll(async () => {
  await browser.close();
});

test("a failing test", async () => {
  await page.goto("https://www.google.com");
  await page.click("text=Get Started fail", { timeout: 1000 }); // << deliberate fail
});

```

Still no luck, I can see the videos are created but the videos are not being embedded in the HTML report which kind makes it difficult to trace which video belong to which result.

![2022-06-13_17-14-25](https://user-images.githubusercontent.com/50574915/173300000-c17c87ae-a053-4e7f-8f85-a724dc869347.png)

![2022-06-13_17-18-02](https://user-images.githubusercontent.com/50574915/173300158-5db1e484-93a4-4eb8-8dc7-815b06decf53.png)

**Comment by dgozman** [comment]
This is indeed a known issue: video recording produces a single video file from the creation of the page until its closure, as of Jun 2022.

**Comment by raulparada** [comment]
Hi!

I'd expect this to be a higher prio bug ([first report September 2021](https://github.com/microsoft/playwright/issues/9002), almost 🍰 🕯️ 🎉 )

I'm following the simple docs on [overriding the page fixture](https://playwright.dev/docs/test-fixtures#overriding-fixtures), **just as described**, and I lose videos because of that? Since not fixing this, there should be a disclaimer on those docs imo

**Comment by mweidner037** [comment]
I'm also running into this issue - specifically, that video recording doesn't work in pages created with `brower.newContext`, as described [here](https://github.com/microsoft/playwright/issues/9002#issuecomment-922187532).

My use case: I'm using `storageState` for authentication, as described in [the docs](https://playwright.dev/docs/test-auth#reuse-signed-in-state), except that I only want to use it for some tests. So when I need it, I call `browser.newContext({ storageState: ... })`. But then pages created from that context don't get video recordings.

For now I plan to use the `recordVideo` option and manual file manipulation to emulate `video: 'retain-on-failure'`.

**Comment by KeionneDerousselle** [comment]
I'm having the same issue at @mweidner037. I'm trying to take retain videos taking on failure, but I'm not seeing any video recordings when my tests fail.

**Comment by NgocVo3112** [comment]
I also get this issue, could you please spend time to fix this. Thanks a lot.

**Comment by angelo-loria** [comment]
Same issue here

**Comment by spiderim** [comment]
I got this issue and tried all the possible methods but could not get the video. Please fix this issue

**Comment by angelo-loria** [comment]
I believe I have this working with a fixture but it's not pretty. A worker-scoped fixture initiates video recording to a temp dir and clean up of that dir on teardown. A test-scoped fixture attaches the video to the test reports on failure or time out.

fixture:
```
import { test as base, BrowserContext } from "@playwright/test";
import * as fs from "fs";

export const test = base.extend<{
    recordVideo: BrowserContext;
    attachVideo: BrowserContext;
}>({
    recordVideo: [
        async ({ browser }, use) => {
            const context = await browser.newContext({
                recordVideo: {
                    dir: "./playwright-video",
                },
            });

            await use(context);

            fs.rmSync("./playwright-video", { recursive: true });
        },
        // @ts-expect-error
        { scope: "worker" },
    ],

    attachVideo: [
        async ({ recordVideo }, use, testInfo) => {
            const path = await recordVideo.pages()[0].video()?.path();

            await use(recordVideo);

            if (
                testInfo.status === "failed" ||
                testInfo.status === "timedOut"
            ) {
                await recordVideo.close();
                testInfo.attach("video", {
                    path: path,
                });
            }
        },
        { scope: "test", auto: true },
    ],
});

export { expect, Page, Locator, Response } from "@playwright/test";

```

usage:
```
import { test, expect, Page } from "../../video-fixture";
import { PageUnderTest } from "../../pageObjects/PageUnderTest"

let page: Page;
let pageUnderTest: PageUnderTest;

test.describe.serial("manage round", () => {
    test.beforeAll(async ({ recordVideo }) => {
        page = await recordVideo.newPage();

        pageUnderTest = new PageUnderTest(page);
        await pageUnderTest.open();
    });

    test.afterAll(async () => {
        await page.close();
    });

    test("renders", async () => {
        await expect(pageUnderTest.header).toHaveTest("header");
        ...
    });

   test("test other stuff on same page", async () => {
        ...
   });
});
```

**Comment by lsnmarcint** [comment]
I have the same problem with reusing StorageState as @mweidner037 and @KeionneDerousselle described.

**Comment by sc0ttdav3y** [comment]
Just adding to this, I'm getting this on MacOS with Playwright version 1.29.1, so it's not just a windows OS issue.

**Comment by vishal7788** [comment]
Hi, I'm also facing the same issue with the Page object Model pattern, the video recording does not work. Waiting for the solution.


`import { test, expect, Page} from "@playwright/test";
import { LoginPage } from "../pages/loginPage";

let page: Page;
let loginPage: LoginPage;

test.describe('Validate login page', async () => {

  test.beforeAll(async ({ browser }) => {
    page = await browser.newPage();
    loginPage = new LoginPage(page);
  
    await loginPage.navigateTo();
  });


  test('login page displayed', async () => {
    await expect(loginPage.Title).toHaveText('ABC');
    await expect(loginPage.Title).toContainText('Log in');
  });

  test('test A', async () => {
    expect('A).toEqual(testA);
  });

  test('testB', async () => {
    expect('B).toEqual(testB);
  });


 test.afterAll(async () => {
    console.log("Login page validated successfully");
    await page.close();
  });
`

**Comment by rubanenkos** [comment]
I have the very same situation - no video attached :(
I've also tried to use the tip from @ryanrosello-og
_contextOptions: {recordVideo: { dir: "<path_to_store_videos>/videos/'"}}_
and 
_const context = await browser.newContext({
    recordVideo: {
      dir: `./videos/`,
    },
  });_
But as a result I've got only empty files in ./videos directory - not video at all

**Comment by daniavander** [comment]
si this is the same issue?
https://github.com/microsoft/playwright/discussions/20946

**Comment by nudratuppal** [comment]
Any plan to fix this issue? 
I am creating new browser contexts by using _storageState_ files for different user roles in fixtures. In this case, videos are not getting created.

**Comment by rishoo2019** [comment]
I am having a similar issue while working with browserContext. 
I hope this gets fixed soon, as this is a critical issue.

**Comment by itz4blitz** [comment]
Our team is also affected by this issue.

**Comment by angelo-loria** [comment]
I've been using the solution I outline in my[ previous comment ](https://github.com/microsoft/playwright/issues/14813#issuecomment-1347264863)for months; it's still working well. I did end up layering another fixture on top of that one for standard page object use.

Fixture for enabling video recording:
```
import { test as base, BrowserContext } from "@playwright/test";
import * as fs from "fs";

export const test = base.extend<{
    recordVideo: BrowserContext;
    attachVideo: BrowserContext;
}>({
    recordVideo: [
        async ({ browser }, use) => {
            const context = await browser.newContext({
                recordVideo: {
                    dir: "./playwright-video",
                },
            });

            await use(context);

            fs.rmSync("./playwright-video", { recursive: true });
        },
        // @ts-expect-error
        { scope: "worker" },
    ],

    attachVideo: [
        async ({ recordVideo }, use, testInfo) => {
            const path = await recordVideo.pages()[0].video()?.path();

            await use(recordVideo);

            if (
                testInfo.status === "failed" ||
                testInfo.status === "timedOut"
            ) {
                await recordVideo.close();
                testInfo.attach("video", {
                    path: path,
                });
            }
        },
        { scope: "test", auto: true },
    ],
});

export { expect, Page, Locator, Response } from "@playwright/test";
```
Then I have my page object fixture:
```
import { test as base, Page } from [fixture for recording video];
import { BrowserContext } from "@playwright/test";
import { ExamplePage1 } from "./examplePage1";
import { ExamplePage2 } from "./examplePage2";
import { ExamplePage3 } from "./examplePage3";

export type PageObjects = {
    examplePage1: ExamplePage1;
    examplePage2: ExamplePage2;
    examplePage3: ExamplePage3;
};

async function getPageContext(recordVideo: BrowserContext): Promise<Page> {
    let page;
    // if page exists use it, else create new one
    if (recordVideo.pages().length === 1) {
        page = recordVideo.pages()[0];
    } else {
        page = await recordVideo.newPage();
    }
    return page;
}

export const test = base.extend<PageObjects>({
    examplePage1: async ({ recordVideo }, use) => {
        const examplePage1 = new ExamplePage1(
            await getPageContext(recordVideo)
        );
        await use(examplePage1);
    },
    examplePage2: async ({ recordVideo }, use) => {
        const examplePage2 = new ExamplePage2(
            await getPageContext(recordVideo)
        );
        await use(examplePage2);
    },
    examplePage3: async ({ recordVideo }, use) => {
        const examplePage3 = new ExamplePage3(
            await getPageContext(recordVideo)
        );
        await use(examplePage3);
    },
});

export { expect, Page, Locator, Response } from "@playwright/test";
```

Usage in spec file:
```
import { test, expect } from [page object fixture above];

// note that a beforeAll() this is required for the video recording to work
test.beforeAll(async ({ examplePage1 }) => {
    await examplePage1.open();
});

test("renders", async ({ examplePage1 }) => {
    await expect(examplePage1.heading).toHaveText(
       "Example Page 1"
    );
});

```

**Comment by angelo-loria** [comment]
> > I've been using the solution I outlined in my[ previous comment ](https://github.com/microsoft/playwright/issues/14813#issuecomment-1347264863)for months; it's still working well. I did end up layering another fixture on top of that one for standard page object use.
> 
> Hi @angelo-loria I'm trying to understand your implementation a bit better for our use case. I had a question I'm hoping you can help with:
> 
> I notice you're using two fixtures with the second being the page object fixture. In some of our tests, the only fixture we are using is browser. Is it possible to extend the browser fixture so that anytime a browser is created and used in a test, a recording is also generated?

Not sure if I'm understanding you correctly but I have the browser context being created in the `recordVideo` fixture. If you're not using any page objects/don't need any other fixtures, you can use the `recordVideo` fixture directly in your spec file in the `beforeAll` hook, like the example in my original comment.

**Comment by shivghai** [comment]
this is my workaround for now in case it helps anyone. not too ugly

```
import { test as base, Page, Video } from '@playwright/test'

import { HomePage } from './../pages/homePage'

/** HomePage.create looks like this

  static async create(browser: Browser, fileName): Promise<Page> {
    if (process.env.BASE_URL == null) {
      throw new Error('No BASE_URL configured')
    }

    const userContext = await browser.newContext({
      storageState: `./e2e-tests/sessionStorage/${fileName}.json`,
      recordVideo:
        {
        dir: `test-results/videos`,
        }
    })
    const userPage = await userContext.newPage()
    await userPage.goto(process.env.BASE_URL)

    return userPage
  }
*/

type Fixtures = {
  userOnePage: Page
  userTwoPage: Page
}
export const test = base.extend<Fixtures & { saveVideos: void }>({
  userOnePage: async ({ browser }, use) => {
    await use(await HomePage.create(browser, 'userOneStorageState'))
  },
  userTwoPage: async ({ browser }, use) => {
    await use(await HomePage.create(browser, 'userTwoStorageState'))
  },
  saveVideos: [async ({ userOnePage, userTwoPage, browser }, use, testInfo) => {
    await use();

    await Promise.all(browser.contexts().map((context) => {
      return context.close()
    }))
    const video1 = await userOnePage.video()?.path()
    const video2 = await userTwoPage.video()?.path()
    const nonNullVideos = [{ title: 'User One', path: video1 }, { title: 'User Two', path: video2 }].filter((video) => Boolean(video.path)) as {title: string, path: string}[]
    testInfo.annotations.push({
      'type': 'Test Evidence Videos',
      'description': 'Test evidence videos are available at the bottom of the test report.',
    })
    nonNullVideos.forEach((video) => {
      testInfo.attachments.push({ name: `Test Evidence: ${video.title} `, path: video.path, contentType: 'video/webm'})
    })
  }, { scope: 'test', auto: true }],

})

```

**Comment by mstepin** [comment]
I think this problem is also present when using worker scoped fixtures. We get no video captures even if we use a new page instance for each test. Would be nice to have a fix for this.

**Comment by david-m-globant** [comment]
I confirm that this bug still occurs in type script, in python the video is saved correctly.

**Comment by daniavander** [comment]
when will this fix develop to the release?

**Comment by Vaishnavi-Thiyagaraj** [comment]
I also get this issue, when will this fix get released?

**Comment by Derrbal** [comment]
We discovered a potential solution to allow us to create a video recording in a test which is using the browser fixture:

**Comment by spaniardmaximus** [comment]
Caring is sharing @Derrbal :). What's the solution?

**Comment by Derrbal** [comment]
> Caring is sharing @Derrbal :). What's the solution?

My apology, I've updated comment ;)

**Comment by ibrocodes7** [comment]
Hi @angelo-loria I wanted to reach out because I'm running into an error while trying out the solution you provided. Any thoughts on why this might be happening? On top of that, I'm also getting an error dialog on the recordVideo object. Your help would be much appreciated!

```ts
    TypeError: Cannot read properties of undefined (reading 'video')
```

<img width="1002" alt="Screenshot 2023-08-04 at 18 04 41" src="https://github.com/microsoft/playwright/assets/65076139/d579c707-80a4-42d7-8257-56b129c01bf5">

Error on recordVideo:
```ts
Type '[({ browser }: { recordVideo: BrowserContext; attachVideo: BrowserContext; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>) => Promise<...>, { ...; }]' is not assignable to type 'TestFixtureValue<BrowserContext, { recordVideo: BrowserContext; attachVideo: BrowserContext; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions> | [...]'.
  Type '[({ browser }: { recordVideo: BrowserContext; attachVideo: BrowserContext; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>) => Promise<...>, { ...; }]' is not assignable to type 'TestFixture<BrowserContext, { recordVideo: BrowserContext; attachVideo: BrowserContext; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions> | [...]'.
    Type '[({ browser }: { recordVideo: BrowserContext; attachVideo: BrowserContext; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>) => Promise<...>, { ...; }]' is not assignable to type '[TestFixtureValue<BrowserContext, { recordVideo: BrowserContext; attachVideo: BrowserContext; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions>, { ...; }]'.
      Type at position 1 in source is not compatible with type at position 1 in target.
        The types of 'scope' are incompatible between these types.
          Type '"worker"' is not assignable to type '"test"'.ts(2322)
video-fixtures.ts(5, 3): The expected type comes from property 'recordVideo' which is declared here on type 'Fixtures<{ recordVideo: BrowserContext; attachVideo: BrowserContext; }, {}, PlaywrightTestArgs & PlaywrightTestOptions, PlaywrightWorkerArgs & PlaywrightWorkerOptions>'
```

**Comment by angelo-loria** [comment]
@ibrocodes7 This is hacky but I would just throw a `// @ts-expect-error` above `recordVideo`. I should revisit this soon and see if there's a cleaner way to implement it. It is still working well for me; I have the fixture in an npm package that is used by a handful of Playwright projects daily and videos are still getting attached to my reports.

**Comment by ibrocodes7** [comment]
@angelo-loria  That's fantastic! I'd love to check out your npm package if its public. It sounds like it could be really beneficial for my projects.

**Comment by angelo-loria** [comment]
@ibrocodes7 it'll do that if you don't have a `beforeAll()` hook in your spec file. I use `beforeAll()` in every spec so it's not been a problem for me.  Might be able to figure out how to have it work without that. The npm package is for my day job/not public but it contains fixtures, a base `PlaywrightTestConfig` object, a global setup function, a custom reporter, and base page objects/page components that are shared across various projects.

**Comment by joeyorlando** [comment]
this is an extension of @Derrbal's comment. This gets rid of the need to copy/paste the video saving logic within individual tests, by use of fixtures:

```ts
import * as fs from 'fs';
import * as path from 'path';
import { test as base, Browser, Page, TestInfo } from '@playwright/test';

import { GRAFANA_ADMIN_USERNAME, GRAFANA_EDITOR_USERNAME, GRAFANA_VIEWER_USERNAME } from './utils/constants';
import { VIEWER_USER_STORAGE_STATE, EDITOR_USER_STORAGE_STATE, ADMIN_USER_STORAGE_STATE } from '../playwright.config';

export class BaseRolePage {
  page: Page;
  userName: string;

  constructor(page: Page) {
    this.page = page;
  }
}

type BaseRolePageType = new (page: Page) => BaseRolePage;

class ViewerRolePage extends BaseRolePage {
  userName = GRAFANA_VIEWER_USERNAME;
}

class EditorRolePage extends BaseRolePage {
  userName = GRAFANA_EDITOR_USERNAME;
}

class AdminRolePage extends BaseRolePage {
  userName = GRAFANA_ADMIN_USERNAME;
}

type Fixtures = {
  viewerRolePage: ViewerRolePage;
  editorRolePage: EditorRolePage;
  adminRolePage: AdminRolePage;
};


const _recordTestVideo = async (
  browser: Browser,
  use: (r: BaseRolePage) => Promise<void>,
  testInfo: TestInfo,
  storageStateLocation: string,
  RolePage: BaseRolePageType
) => {
  const videoDir = path.join(testInfo.outputPath(), 'videos');

  const context = await browser.newContext({ storageState: storageStateLocation, recordVideo: { dir: videoDir } });
  const page = new RolePage(await context.newPage());

  try {
    await use(page);
  } finally {
    await context.close();
    const videoFiles = fs.readdirSync(videoDir);

    if (videoFiles.length > 0) {
      for (let i = videoFiles.length; i > 0; i--) {
        let videoFile = path.join(videoDir, videoFiles[i - 1]);
        await testInfo.attach('video', { path: videoFile });
      }
    }
  }
};

export * from '@playwright/test';
export const test = base.extend<Fixtures>({
  viewerRolePage: ({ browser }, use, testInfo) =>
    _recordTestVideo(browser, use, testInfo, VIEWER_USER_STORAGE_STATE, ViewerRolePage),
  editorRolePage: async ({ browser }, use, testInfo) =>
    _recordTestVideo(browser, use, testInfo, EDITOR_USER_STORAGE_STATE, EditorRolePage),
  adminRolePage: async ({ browser }, use, testInfo) =>
    _recordTestVideo(browser, use, testInfo, ADMIN_USER_STORAGE_STATE, AdminRolePage),
});
```

then within your test files you can do something like this:
```ts
import { test, expect, Page } from '../fixtures';

test('my test', async ({ adminRolePage }) => {
  ...
```

The `adminRolePage` fixture that is accessed in `test` in this example will automatically have a video recorded.

**Comment by asaf-shochet** [comment]
Same issue here, using latest Playwright version `1.38.0`.

We are testing a Chrome Extension, following the official tutorial, all works great except for the videos.
Following the advice here, I added the `contextOptions` option to `use` on the playwright.config.ts file
```
use: {
	video: 'retain-on-failure',
	contextOptions: { recordVideo: { dir: 'test-results/videos/' } }
	},
```

I see the videos generated, but they are not connected to the HTML report, or the JUnit XML format.

**Comment by zacharya19** [comment]
If you are like me and have long running tests, from some reason the video is not yet flushed before getting to `fs.readdirSync`, which causes an error and partial video file.
My current workaround is to add sleep of 2 sec bfore this function.

> We discovered a potential solution to allow us to create a video recording in a test which is using the browser fixture:
> 
> ```ts
> import * as path from "path"
> import * as fs from 'fs';
> import { expect, test } from "@playwright/test";
> 
> test.only('manually created context', async ({ browser }, testInfo) => {
>   const videoDir = path.join(testInfo.outputPath(), 'videos');
>   const context = await browser.newContext({ recordVideo: { dir: videoDir } });
>   const page = await context.newPage();
>   const page2 = await context.newPage();
>   try {
>     await page2.goto('http://localhost:3030/login?pageSize=10');
>     await page.goto('http://localhost:3030/login?pageSize=10');
>     await expect(page).toHaveURL('http://localhost:3030/login?pageSize=10');
>   } finally {
>     await page.context().close();
>     const videoFiles = fs.readdirSync(videoDir);
>     
>     if (videoFiles.length > 0) {
>       for (let i = videoFiles.length; i > 0; i--) {
>         let videoFile = path.join(videoDir, videoFiles[i - 1]);
>         await testInfo.attach('video', { path: videoFile });
>       }
>     }
>   }
> });
> ```
> 
> But adding this change accross all of our tests which uses the browser fixture is tome consuming and not ideal. We are hoping that someone has some ideas how we might abstract this change. We currently have extended test to use custom fixtures. It seems that it can be a good place to globally apply this fix but we are not sure how.

**Comment by zacharya19** [comment]
If you are like me and have long running tests, from some reason the video is not yet flushed before getting to `fs.readdirSync`, which causes an error and partial video file (`Error: file data stream has unexpected number of bytes`).
My current workaround is to add sleep of 2 sec bfore this function.

> We discovered a potential solution to allow us to create a video recording in a test which is using the browser fixture:
> 
> ```ts
> import * as path from "path"
> import * as fs from 'fs';
> import { expect, test } from "@playwright/test";
> 
> test.only('manually created context', async ({ browser }, testInfo) => {
>   const videoDir = path.join(testInfo.outputPath(), 'videos');
>   const context = await browser.newContext({ recordVideo: { dir: videoDir } });
>   const page = await context.newPage();
>   const page2 = await context.newPage();
>   try {
>     await page2.goto('http://localhost:3030/login?pageSize=10');
>     await page.goto('http://localhost:3030/login?pageSize=10');
>     await expect(page).toHaveURL('http://localhost:3030/login?pageSize=10');
>   } finally {
>     await page.context().close();
>     const videoFiles = fs.readdirSync(videoDir);
>     
>     if (videoFiles.length > 0) {
>       for (let i = videoFiles.length; i > 0; i--) {
>         let videoFile = path.join(videoDir, videoFiles[i - 1]);
>         await testInfo.attach('video', { path: videoFile });
>       }
>     }
>   }
> });
> ```
> 
> But adding this change accross all of our tests which uses the browser fixture is tome consuming and not ideal. We are hoping that someone has some ideas how we might abstract this change. We currently have extended test to use custom fixtures. It seems that it can be a good place to globally apply this fix but we are not sure how.

**Comment by iburahim786** [comment]
+1

**Comment by drupal-spider** [comment]
My test environment:
npm list
├── @playwright/test@1.39.0
├── @types/node@20.8.9
└── dotenv@16.3.1
OS: MacOS 14.1
Chip: Apple M2 Max
Browser: Chromium Version 119.0.6045.9 (Developer Build) (arm64)

Result:
 Adding the following to playwright.config.ts
```
export default defineConfig({
  use: {
    // Capture screenshot after each test failure.
    screenshot: 'on',
    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'off',
    /* Record the video for test. See https://playwright.dev/docs/videos */
    // Record video only when retrying a test for the first time.
    video: 'on',
    contextOptions: { recordVideo: { dir: 'test-results/videos/' } }
  },
...
});

```
Videos are generated in the output folder 'test-results/videos'. But I have to close the browser for each test, otherwise the video will be only generated once.

**Comment by drupal-spider** [comment]
Updated:

I used VS Code + Playwright Test for VSCode extension(1.0.17) to run the test. In the 'Testing' tab, if I check the Show browser box, the video won't be generated. But if I tick the box 'Show trace viewer', the video was generated and attached to the report.

**Comment by 666f78** [comment]
Same issue, using latest Playwright version 1.40.0.

**Comment by ryanrosello-og** [comment]
hey @dgozman , any plans on fixing this?  Seems users are still hitting this issue.  Some have proposed workarounds.

**Comment by o-holyshevskyi** [comment]
Same issue here on the latest version 1.40.0

**Comment by Varsha176** [comment]
Facing same issue with  latest version 1.40.0. Video not getting generated .

**Comment by magarwal19** [comment]
Hi @dgozman 
can you please comment if we can get it resolved in 1.41 version?

we have folded other issue (https://github.com/microsoft/playwright/issues/28327) in this and is important for us to get it resolved

**Comment by magarwal19** [comment]
@mxschmitt 
can you please help us to prioritise it?

**Comment by margarytaD** [comment]
Also have this issue with re-using storageState

**Comment by magarwal19** [comment]
@dgozman 
can you please help us to know if this is in road map for near future?

**Comment by mstepin** [comment]
Whichever workaround that one uses, the result is less than ideal as the video sits in the attachments section of the HTML report, instead of the "Videos" section (which I note uses the **plural** form of the word, but have not ever seen more than a single video in that section, no matter how I play around with browsers, contexts, or pages)

**Comment by vladkrasn** [comment]
This issue is a concern for me too!

**Comment by angelo-loria** [comment]
> Whichever workaround that one uses, the result is less than ideal as the video sits in the attachments section of the HTML report, instead of the "Videos" section (which I note uses the **plural** form of the word, but have not ever seen more than a single video in that section, no matter how I play around with browsers, contexts, or pages)

That is not true, videos are properly attached to the HTML report under the Videos section with the fixtures [I posted earlier in this thread.](https://github.com/microsoft/playwright/issues/14813#issuecomment-1489363592)

**Comment by siddharth2023** [comment]
bump-up

**Comment by MaddoCXX** [comment]
Any updated on fixes on this one?  I still encountering issue. see below code


import { defineConfig, devices } from '@playwright/test';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */
export default defineConfig({
  testDir: './tests',
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: 'html',
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'],
      viewport: {width:1920, height:1080},
      screenshot: "on",
      video: "on",
      trace:"on"
      
    },
    },

    // {
    //   name: 'firefox',
    //   use: { ...devices['Desktop Firefox'] },
    // },

    // {
    //   name: 'webkit',
    //   use: { ...devices['Desktop Safari'] },
    // },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: { ...devices['Pixel 5'] },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: { ...devices['iPhone 12'] },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },
    // },
    // {
    //   name: 'Google Chrome',
    //   use: { ...devices['Desktop Chrome'], channel: 'chrome' },
    // },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },
});

![image](https://github.com/microsoft/playwright/assets/112743692/ebae0062-3122-46f2-acd7-3bd713a58902)

**Comment by shubnit12** [comment]
Videos are not Generated in playwright-report html report

Below the playwright.config.ts file
```
import { defineConfig, devices } from '@playwright/test';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */
export default defineConfig({
  testDir: './tests',
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: 'html',
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',
    headless: true,
    screenshot: "only-on-failure",
    video: "on"
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
    },

    // {
    //   name: 'firefox',
    //   use: { ...devices['Desktop Firefox'] },
    // },

    // {
    //   name: 'webkit',
    //   use: { ...devices['Desktop Safari'] },
    // },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: { ...devices['Pixel 5'] },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: { ...devices['iPhone 12'] },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },
    // },
    // {
    //   name: 'Google Chrome',
    //   use: { ...devices['Desktop Chrome'], channel: 'chrome' },
    // },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },
});

```
Below is the test

```
import { chromium, BrowserContext } from 'playwright';
import { test } from '@playwright/test';

test('testxx', async () => {
    const browser = await chromium.launch();
    const context: BrowserContext = await browser.newContext();
    const page = await context.newPage();
    await page.goto('https://www.google.com');
    await context.close();
    await browser.close();
    
});
```
Attached Screenshot

![image](https://github.com/microsoft/playwright/assets/84227718/51918295-8204-45c7-9483-da3a4b91aab1)

**Comment by graziano-mck** [comment]
> This is indeed a known issue: video recording produces a single video file from the creation of the page until its closure, as of Jun 2022.

Hey @dgozman, even though I follow what has been suggested [here](https://playwright.dev/docs/videos#record-video)
```
Videos are saved upon [browser context](https://playwright.dev/docs/browser-contexts) closure at the end of a test. If you create a browser context manually, make sure to await [browserContext.close()](https://playwright.dev/docs/api/class-browsercontext#browser-context-close).
```
I indeed await `browserContext.close()`, but there are no videos created.

**Comment by Akallabet** [comment]
> This is indeed a known issue: video recording produces a single video file from the creation of the page until its closure, as of Jun 2022.

Hey @dgozman, even though I follow what has been suggested [here](https://playwright.dev/docs/videos#record-video)

> Videos are saved upon [browser context](https://playwright.dev/docs/browser-contexts) closure at the end of a test.
If you create a browser context manually, make sure to await [browserContext.close()](https://playwright.dev/docs/api/class-browsercontext#browser-context-close).

I indeed await `browserContext.close()`, but there are no videos created.

**Comment by poponuts** [comment]
Almost 2 years in and still an issue. I think this is very useful for debugging locally. I hope the Playwright core team starts looking into this with high priority 😄

**Comment by ollyjohn** [comment]
Hey @angelo-loria, I'm trying to use the first example you posted as we're not exposing our POMs via fixtures but I'm unable to get the path by calling `.path()` on the video per your example as I'm connecting remotely.  The error thrown suggests using `saveAs` and providing the path myself, but this just seems to hang indefinitely; appreciate this isn't necessarily related, but just wondered if you've run into this before at all and have any sage advice to offer?

**Comment by jon-ressio** [comment]
It was mentioned above but this is an issue not just for custom fixtures, but overriding storageState too.  I'd really rather avoid having to write a custom fixture to get around this since I only need the storageState override for auth.

**Comment by michael-hoss** [comment]
Is the only trigger to actually save the videos to disk actually `context.close()`?

I want to export individual videos for each test within a `describe` environment, and re-use the context and page across the tests. 

However, when I call 
- `recordVideoContext.pages()[0].video()?.saveAs(videoFilenameForThisTest);` in a fixture after each test
- and `await recordVideoContext.close();` in the `afterAll` hook

then all the videos are essentially the same, and they all contain the recording of the entire `describe` environment :thinking: 

I guess this is just yet another finding of the original problem that @dgozman confirmed ([link](https://github.com/microsoft/playwright/issues/14813#issuecomment-1154579148)) at the beginning of this issue thread

> This is indeed a known issue: video recording produces a single video file from the creation of the page until its closure, as of Jun 2022.

**Comment by alex-vungle** [comment]
It's still reproducible on Playwright v1.42.1

**Comment by moorthyptm** [comment]
@alex-vungle its reproducible https://playwright.dev/docs/auth#testing-multiple-roles-with-pom-fixtures

try: Testing multiple roles with POM fixtures

this case we are creating new browser context 

` 
//  Admin fixture
export const test = base.extend<{adminPage1: adminPage1, adminPage2: adminPage2}>({
    adminPage1: async ({page}, use) => {
        const context = await browser.newContext({
            storageState: wallet.loginAdmin,
        });
        //   const page = await context.newPage();
        const adminPage1 = new adminPage1(context.newPage());
        const adminPage2 = new adminPage2(context.newPage());
        await use({adminPage1, adminPage2}); // Even i have tried to use single adminPage but still recording is not happening [example await use(adminPage1)]
        await context.close();
}});



// Expected Test execution with recording
// result : No recording
test('test', async ({adminPage}) => {

    // Test code here 
    
}

// Experiment Test [accessing page prop]
// Result: Recording working but its empty page [since iam not using page prop for test instead using adminPage]
test('test', async ({adminPage,page}) => {
    console.log(page)
    // Test code here 
    
}`


using version :  "@playwright/test": "^1.44.1",

**Comment by moorthyptm** [comment]
While creating a new context, if I provide a video directory, it records the video, but I want the video to be attached to the test result.

 const context = await browser.newContext({
    recordVideo: {
      dir: `./videos/`,
    },
  });

**Comment by moorthyptm** [comment]
This resource helped me resolve the issue: [Playwright is not generating videos when we create our own context](https://stackoverflow.com/questions/78170953/playwright-is-not-generating-videos-when-we-create-our-own-context).


https://stackoverflow.com/questions/78170953/playwright-is-not-generating-videos-when-we-create-our-own-context


 With the help of the test info, I was able to attach the video to the report.

**Comment by Romarionijim** [comment]
Im having issues in both modes, 'retain-on-failure' and 'on'.

**Comment by Romarionijim** [comment]
@moorthyptm could this be done in the config file somehow instead of each test file? I have above 600 tests so repeating this steps over and over seems an overkill. 

or could be exported as a fixture.

**Comment by moorthyptm** [comment]
Hey @Romarionijim 

For reusability try https://playwright.dev/docs/test-fixtures fixtures

**Comment by magarwal19** [comment]
Is there any solution which can be done using config file?
also my code has both type of test
1. Which use default context (getting videos here captured correctly)
2. Which use multiple context (not getting videos here)

If i update the report directory in config file, then it does capture the videos, but then not tagged to HTML file and is not helpful overall as it becomes hard to distinguish videos and map to tests

**Comment by MicleFilimonov** [comment]
Guys. maybe it will help for someone of you 
i'm receiving all videos for html reporter after running my test - for my it neccesary to run one test over and over with changing variable and i'am prepearing "skeleton" for this test. So after each test i've got my videos to the reporter on failure state

So this is my config 
```
// @ts-nocheck
const { defineConfig, devices } = require('@playwright/test');

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config({ path: path.resolve(__dirname, '.env') });

/**
 * @see https://playwright.dev/docs/test-configuration
 */
module.exports = defineConfig({
  testDir: './tests',
  /* Run tests in files in parallel */
  fullyParallel: false,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [
    ["html"], 
    ["allure-playwright"]
  ],
  
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /*video, screenshot, headlessmode  */
    video: 'retain-on-failure',
    screenshot: 'only-on-failure',
    headless: false,
    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'retain-on-failure',
    
    // launchOptions: {},
  },

  /* Configure projects for major browsers */
  projects: [
    // {
    //   name: 'chromium',
    //   use: { ...devices['Desktop Chrome'] },
    // },

    // {
    //   name: 'firefox',
    //   use: { ...devices['Desktop Firefox'] },
    // },

    // {
    //   name: 'webkit',
    //   use: { ...devices['Desktop Safari'] },
    // },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: { ...devices['Pixel 5'] },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: { ...devices['iPhone 12'] },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },
    // },
    {
      name: 'Google Chrome',
      use: { ...devices['Desktop Chrome'], channel: 'chrome' },
    },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },

});

```


and this is my test wrapped into for cycle for multiple executions 
```
const { test, expect, chromium } = require('@playwright/test');

for (let i = 0; i < 2; i++) {
  test(`Checking site - run #${i + 1}`, async ({ page }) => {
  
    await test.step('Открытие страницы входа', async () => {
      await page.setViewportSize({ width: 1920, height: 1080 });
      await page.goto('https://example.com');
    });

    await test.step('Ввод email и пароля', async () => {
      await page.waitForSelector('text=Email');
      await page.locator('#email').fill('example@email.com');
      await page.locator('#password').fill('123456');
    });

    await test.step('Клик на кнопку входа', async () => {
      await page.locator('#sign_in').click();
    });

    await test.step('Проверка видимости иконки уведомлений', async () => {
      const Bell = page.locator('#bell');
      await Bell.waitFor({ state: 'visible', timeout: 2000 });
      await expect(Bell).toBeVisible();

      await test.step('Закрытие текущего браузера', async () => {
        await browser.close();
      });
    });
  })
}
```


for this multiple runs (2 or whatever you want times) i'am getting video for every test execution in html reporter (so and in to allure reporter too if you desire) 

![image](https://github.com/user-attachments/assets/aaf9cad7-c592-4815-894c-5af251c52673)

I was read all your comments and try to solve this problem, so i'm hoping this solutions will help to someone here

**Comment by RussellJQA** [comment]
I'm learning playwright, and (while learning) wanted to (for now) always record a test video.

I had a difficult time getting that working. But I finally found a workaround that I wanted to share, in the hope that it's somehow helpful for someone else here.

Based on some of what I'd seen above, I'd initially tried various options in the "use" section of my playwright.config.ts file, including (but not limited to):

    // video: "on", // That didn't save the videos anywhere that I could find.

    // contextOptions: {recordVideo: {dir: './playwright-report/data'}}, // That **temporarily** saved the videos: But they videos were deleted when the tests were complete.

    // contextOptions: {recordVideo: {dir: './test-results'}}, // That saved the videos: But they weren't included in my tests' HTML report.

None of those options "worked". [None of them (lastingly) saved the test videos AND included them in my tests' HTML report.]

Finally, instead of trying to do this in my playwright.config.ts file, I tried doing it within a specific "XXX.ts" test file, using "test.use({video: 'on'});".

That "worked":
• For each test in the test file, it saved a "video.webm" in that test's subfolder (of the default "test-results" folder)
• The videos were all included in my test's HTML report (in './playwright-report/index.html'), which I accessed via 'npx playwright show-report'
• This was the only option I found which logged a stringified "video" fixture (the default version) as "on" rather than as "off"

I've attached a minimalistic (though not minimal) example of what I did. It's a "working" modified version (Zipped) of the example.spec.ts file which gets automatically installed with the TS version of Playwright.

Environment:
• Language: TypeScript (node v20.17.0, npm 10.8.2)
• playwright Version 1.46.1
• Browser: playwright's chromium browser [Chromium Version 128.0.6613.18 (Developer Build) (64-bit)]
• OS: Windows 11 Professional (Version: 23H2, OS build: 22631.4037)
[record videos.test.zip](https://github.com/user-attachments/files/16825740/record.videos.test.zip)

**Comment by caestrada1103** [comment]
Guys I have solved this problem in another way, this would be useful for someone who needs something like I did

Here is a little more context:

As many know the Playwright documentation is a bit ambiguous, and doesn't show much in the way of examples that might be similar to reality.

In my case, I needed to enable videos only when a test case failed, such as video: 'retain-on-failure'.

What happened, when Parallelism is implemented, it somehow makes the generic use of playwright.config.ts, unusable if you are using customFixtures, this makes it overwrite and retain some settings.

After solving the parallelism problem with multiple workers and different customFixtures to avoid crossing information, I found that it was necessary to create a new context (suggested by the documentation), this new context makes that the videos are not recorded and it is necessary to pass the configuration manually (since a new context is a manual context that we create on demand).

My solution was the following, keeping different contexts, enabling parallelism to keep the recordings when there is a failure (for debug or other purpose):

```
type testFixture = {
    page: Page;
    request: APIRequestContext;
};

/**
 * A Playwright test fixture for tests where the admin role performs most of the actions.
 *
 * This fixture extends the base test with additional setup specific to admin users.
 * It navigates to an admin page and configures
 * API request headers for the admin role.
 *
 * @typeParam testFixture - The type of the fixture used in the test context.
 *
 * @property page - A Playwright `page` object. It navigates to the admin login page,
 *                  and can provides the logged-in page instance for testing.
 * @property request - A Playwright `request` object. It generates admin-specific
 *                     authentication headers.
 */
export const adminTest = base.extend<testFixture>({
    page: async ({ browser }, use, testInfo) => {
        //Creation of new context of browser to avoid overlap data and behaviors, adding recordVideo property of context,
       //That has the relative path for videos of each execution
        const context = await browser.newContext({
            recordVideo: {
                dir: testInfo.outputPath('videos'),
            },
        });
        //Base on the new context create a new page to interact with
        const page = await context.newPage();
        //doing some configuration, load sessions, setting up database or something
        await page.goto('<your.url.com>');
        await page.waitForLoadState('load');
        await use(page);
        //After all interactions the context needs to be close to close the browser to avoid
        //keep browser (page) open, this is know as a teardown and ONLY works AFTER the use() method
        // Include some conditional base on status test AFTER execution to create the video, e.i when a test fail
        if (
            testInfo.status === 'failed' ||
            testInfo.status === 'timedOut' ||
            testInfo.status === 'interrupted'
        ) {
            //Dynamic path to avoid overlaping videos for multiple workers
            const videoPath = testInfo.outputPath('videos', `${testInfo.testId}.webm`);
            // Save path and close the current context to attatch it into report
            await Promise.all([page.video().saveAs(videoPath), context.close()]);
            // Attatch the video to the report.
            testInfo.attachments.push({
                name: 'video',
                path: videoPath,
                contentType: 'video/webm',
            });
        // In case that the test case pass as expected, just close the context
        } else context.close();
    },
    request: async ({ request }, use) => {
        //Doing some configuration, load tokens, etc
        await use(request);
    },
});
```

**Comment by caestrada1103** [comment]
> > Guys I have solved this problem in another way, this would be useful for someone who needs something like I did
> > Here is a little more context:
> > As many know the Playwright documentation is a bit ambiguous, and doesn't show much in the way of examples that might be similar to reality.
> > In my case, I needed to enable videos only when a test case failed, such as video: 'retain-on-failure'.
> > What happened, when Parallelism is implemented, it somehow makes the generic use of playwright.config.ts, unusable if you are using customFixtures, this makes it overwrite and retain some settings.
> > After solving the parallelism problem with multiple workers and different customFixtures to avoid crossing information, I found that it was necessary to create a new context (suggested by the documentation), this new context makes that the videos are not recorded and it is necessary to pass the configuration manually (since a new context is a manual context that we create on demand).
> > My solution was the following, keeping different contexts, enabling parallelism to keep the recordings when there is a failure (for debug or other purpose):
> > ```
> > type testFixture = {
> >     page: Page;
> >     request: APIRequestContext;
> > };
> > 
> > /**
> >  * A Playwright test fixture for tests where the admin role performs most of the actions.
> >  *
> >  * This fixture extends the base test with additional setup specific to admin users.
> >  * It navigates to an admin page and configures
> >  * API request headers for the admin role.
> >  *
> >  * @typeParam testFixture - The type of the fixture used in the test context.
> >  *
> >  * @property page - A Playwright `page` object. It navigates to the admin login page,
> >  *                  and can provides the logged-in page instance for testing.
> >  * @property request - A Playwright `request` object. It generates admin-specific
> >  *                     authentication headers.
> >  */
> > export const adminTest = base.extend<testFixture>({
> >     page: async ({ browser }, use, testInfo) => {
> >         //Creation of new context of browser to avoid overlap data and behaviors, adding recordVideo property of context,
> >        //That has the relative path for videos of each execution
> >         const context = await browser.newContext({
> >             recordVideo: {
> >                 dir: testInfo.outputPath('videos'),
> >             },
> >         });
> >         //Base on the new context create a new page to interact with
> >         const page = await context.newPage();
> >         //doing some configuration, load sessions, setting up database or something
> >         await page.goto('<your.url.com>');
> >         await page.waitForLoadState('load');
> >         await use(page);
> >         //After all interactions the context needs to be close to close the browser to avoid
> >         //keep browser (page) open, this is know as a teardown and ONLY works AFTER the use() method
> >         // Include some conditional base on status test AFTER execution to create the video, e.i when a test fail
> >         if (
> >             testInfo.status === 'failed' ||
> >             testInfo.status === 'timedOut' ||
> >             testInfo.status === 'interrupted'
> >         ) {
> >             //Dynamic path to avoid overlaping videos for multiple workers
> >             const videoPath = testInfo.outputPath('videos', `${testInfo.testId}.webm`);
> >             // Save path and close the current context to attach it into report
> >             await Promise.all([page.video().saveAs(videoPath), context.close()]);
> >             // Attach the video to the report.
> >             testInfo.attachments.push({
> >                 name: 'video',
> >                 path: videoPath,
> >                 contentType: 'video/webm',
> >             });
> >         // In case that the test case pass as expected, just close the context
> >         } else context.close();
> >     },
> >     request: async ({ request }, use) => {
> >         //Doing some configuration, load tokens, etc
> >         await use(request);
> >     },
> > });
> > ```
> 
> Hey,
> 
> Wow that's cool, it might just be the solution for my case as well!
> 
> Just one quick question, where did you put this setup in your project structure? I mean, it's a separate module, right?
> 
> Much obliged 👍

Thank you!.

This file contain all custom fixtures arround my suite, this is like test from playwright, but insteand of been imported by playwright/test, i imported form this file.

now you can implemented on the test suite like adminTest, using all kind of hooks and layers arround the test, like describe, test itself (just the adminTest in this case) and maybe steps

here is some examples.

```
import { adminTest} from 'support/fixtures/customFixtures';

//hooks for all describes
adminTest.describe('this would be a feature for your test', () => {
    //Some hooks for test cases...
    adminTest.beforeEach('this is a simple hook', async ({ page }) => {
        //Do something
    });
    adminTest(
        'test case description',
        {
            tag: ['@admin', '@smoke'],
        },
        async ({ page }) => {
            //some initialization of variables
            await adminTest.step('this is a step o many steps', async () => {
                page.goto('/');
                //...
                //Some steps to define actions or something
            });
        },
    );
});

```

In case that didn't work, rememeber to add this line as import of the custom fixture...

`import { APIRequestContext, test as base, Page } from '@playwright/test';`

**Comment by tomer-dev** [comment]
Hi there, this behavior breaks any `browser.newContext()` essentially?

> Hi!
> 
> I'd expect this to be a higher prio bug ([first report September 2021](https://github.com/microsoft/playwright/issues/9002), almost 🍰 🕯️ 🎉 )
> 
> I'm following the simple docs on [overriding the page fixture](https://playwright.dev/docs/test-fixtures#overriding-fixtures), **just as described**, and I lose videos because of that? Since not fixing this, there should be a disclaimer on those docs imo

**Comment by ValorHeart** [comment]
Here's how to add a 'mobilePage' fixture that will use mobile resolution and parameters and also record a video:

Use case: Hybrid test that opens Desktop browser page, and then shifts into Mobile browser page  for end to end simulation between interfaces.

Based on @alexzavg 's & Playwright's internal code solutions

```
export const test = base.extend<CustomFixtures & CustomOptions>({
  mobilePage: async ({ browser, contextOptions }, use, testInfo) => {
    const videoMode = testInfo.project.use.video;
    // Whether to start video recording
    const captureVideo = videoMode === 'on' || videoMode === 'retain-on-failure' || (videoMode === 'on-first-retry' && testInfo.retry === 1);

    const mobileContext = await browser.newContext({
      ...contextOptions,
      ...devices['iPhone 15'],
      ...(captureVideo && { recordVideo: { dir: testInfo.outputPath('videos') } }),
    });

    const mobilePage = await mobileContext.newPage();
    await use(mobilePage);

    // Whether to save & attach video after test
    const testFailed = testInfo.status !== testInfo.expectedStatus;
    const preserveVideo =
      captureVideo &&
      (videoMode === 'on' || (testFailed && videoMode === 'retain-on-failure') || (videoMode === 'on-first-retry' && testInfo.retry === 1));
    if (preserveVideo) {
      try {
        const videoPath = testInfo.outputPath('videos', `${testInfo.testId}.webm`);
        await Promise.all([mobilePage.video()?.saveAs(videoPath), mobileContext.close()]);
        testInfo.attachments.push({ name: 'video', path: videoPath, contentType: 'video/webm' });
      } catch (err) {
        console.error('💥 Failed to attached video for mobilePage', err);
      }
    } else {
      await mobileContext.close();
    }
  },
});
```

**Comment by Zelys-DFKH** [comment]
Checked this on 1.62.1 to see whether anything had changed.

The reported behavior still reproduces. With `recordVideo` on a context you create yourself, one page reused across three tests, and the context closed only at the end, you get a single file covering all three: one video, 3.04s. That matches what @dgozman described in 2022. The runner saves and attaches video from the fixture that creates the context and closes it per test, so when you own the context that path never runs.

`page.screencast.start()` / `page.screencast.stop()` landed in v1.59, after this was filed. It can be stopped and started again on a page that stays open, so each test can bracket its own recording. Same setup, one page never closed between tests, produced one file per test.

This fixture works today:

```ts
import { test as base, type Page } from '@playwright/test';

const test = base.extend<{}, { sharedPage: Page }>({
  sharedPage: [async ({ browser }, use) => {
    const context = await browser.newContext();
    await use(await context.newPage());
    await context.close();
  }, { scope: 'worker' }],
});

test.beforeEach(async ({ sharedPage }, testInfo) => {
  await sharedPage.screencast.start({ path: testInfo.outputPath('video.webm') });
});

test.afterEach(async ({ sharedPage }, testInfo) => {
  await sharedPage.screencast.stop();
  await testInfo.attach('video', { path: testInfo.outputPath('video.webm'), contentType: 'video/webm' });
});
```

One shared page across the run, and a valid VP8 WebM attached per test in the report.

`recordVideo` itself is unchanged, so the original report stands. This just gives the reused-page case per-test videos in the meantime.

**Comment by angelo-loria** [comment]
@Zelys-DFKH  you could also replace those `beforeEach` and `afterEach` hooks with a fixture

```
const test = base.extend<{ recordScreencast: void }, { sharedPage: Page }>({
  sharedPage: [
    async ({ browser }, use) => {
      const context = await browser.newContext();
      await use(await context.newPage());
      await context.close();
    },
    { scope: "worker" },
  ],

  recordScreencast: [
    async ({ sharedPage }, use, testInfo) => {
      const fileName = `${testInfo.title.replace(/\s+/g, "_")}.webm`;
      const path = testInfo.outputPath(fileName);
      await sharedPage.screencast.start({ path: path });
      await use();
      await sharedPage.screencast.stop();
      if (testInfo.status !== "passed") {
        await testInfo.attach("video", {
          path: path,
          contentType: "video/webm",
        });
      }
    },
    { scope: "test", auto: true },
  ],
});

```

edit: fix syntax from copying from my own repo to here

**Comment by Zelys-DFKH** [comment]
Nicer, thanks. Two things stop that snippet compiling as written: the type declares `saveScreencast` but the fixture is named `recordScreencast`, and the body destructures `reusablePage` where the worker fixture is `sharedPage`. Your version also only attaches on failure, which is fine as a preference but a behavior change rather than a cleanup.

Corrected, and run on 1.62.1:

```ts
const test = base.extend<{ recordScreencast: void }, { sharedPage: Page }>({
  sharedPage: [async ({ browser }, use) => {
    const context = await browser.newContext();
    await use(await context.newPage());
    await context.close();
  }, { scope: 'worker' }],

  recordScreencast: [async ({ sharedPage }, use, testInfo) => {
    const path = testInfo.outputPath(`${testInfo.title.replace(/\s+/g, '_')}.webm`);
    await sharedPage.screencast.start({ path });
    await use();
    await sharedPage.screencast.stop();
    await testInfo.attach('video', { path, contentType: 'video/webm' });
  }, { scope: 'test', auto: true }],
});
```

Three tests, one page for the whole run, three files: VP8 800x450, 1.2s each, matching each test's own duration. Also confirmed the page is shared and not recreated: set a global in one test, read it back in the next.
